lib/puppet/provider/elb_loadbalancer/v2.rb

```diff
@@ -5,60 +5,114 @@
 
   mk_resource_methods
 
-  def self.instances
+  def self.instances(ref_catalog=nil)
+    Puppet.debug('Fetching ELB instances')
     regions.collect do |region|
-      retries = 0
-      begin
         load_balancers = []
-        region_client = elb_client(region)
-        region_client.describe_load_balancers.each do |response|
-          response.data.load_balancer_descriptions.collect do |lb|
-            load_balancers << new(load_balancer_to_hash(region, lb))
+        elbs do |lb|
+          retries = 0
+          begin
+            load_balancers << new(load_balancer_to_hash(region, lb, ref_catalog))
+          rescue Aws::EC2::Errors::RequestLimitExceeded => e
+            retries += 1
+            if retries <= 8
+              sleep_time = 2 ** retries
+              Puppet.debug("Request limit exceeded, retry in #{sleep_time} seconds")
+              sleep(sleep_time)
+              retry
+            else
+              raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, self.resource_type.name.to_s, e.message)
+            end
+          rescue Timeout::Error, StandardError => e
+            raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, self.resource_type.name.to_s, e.message)
           end
         end
+
         load_balancers
-      rescue Aws::EC2::Errors::RequestLimitExceeded => e
-        retries += 1
-        if retries <= 8
-          sleep_time = 2 ** retries
-          Puppet.debug("Request limit exceeded, retry in #{sleep_time} seconds")
-          sleep(sleep_time)
-          retry
-        else
-          raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, self.resource_type.name.to_s, e.message)
+    end.flatten
+  end
+
+  def self.elbs
+    # Make the calls to elb_client for each of the regions to fetch the ELB
+    # resource information, yielding the individual ELB objects.
+    regions.collect do |region|
+      region_client = elb_client(region)
+      Puppet.debug('Calling for 10 ELBs')
+      response = region_client.describe_load_balancers({
+        page_size: 10
+      })
+      marker = response.next_marker
+
+      response.load_balancer_descriptions.each do |lb|
+        yield lb
+      end
+
+      while marker
+        Puppet.debug('Calling for 10 ELBs')
+        response = region_client.describe_load_balancers({
+          page_size: 10,
+          marker: marker
+        })
+        marker = response.next_marker
+        response.load_balancer_descriptions.each do |lb|
+          yield lb
         end
-      rescue Timeout::Error, StandardError => e
-        raise PuppetX::Puppetlabs::FetchingAWSDataError.new(region, self.resource_type.name.to_s, e.message)
       end
-    end.flatten
+    end
   end
 
   read_only(:region, :scheme, :listeners, :tags, :dns_name)
 
   def self.prefetch(resources)
-    instances.each do |prov|
+    ref_catalog = resources.values.first.respond_to?(:catalog) ? resources.values.first.catalog : nil
```

Comment:
I know this caused a spec failure, but I think the spec may be wrong... I can't think of any time that prefetch would be called in the real world that WOULDN'T have a populated `resources` hash.